statefun-flink/statefun-flink-core/src/main/java/org/apache/flink/statefun/flink/core/message/MessageFactoryKey.java

```diff
@@ -0,0 +1,43 @@
+package org.apache.flink.statefun.flink.core.message;
+
+import java.io.Serializable;
+import java.util.Objects;
+
+public class MessageFactoryKey implements Serializable {
+  private static final long serialVersionUID = 1L;
+
+  private final MessageFactoryType type;
+  private final String customPayloadSerializerClassName;
+
+  private MessageFactoryKey(MessageFactoryType type, String customPayloadSerializerClassName) {
+    this.type = type;
+    this.customPayloadSerializerClassName = customPayloadSerializerClassName;
+  }
+
+  public static MessageFactoryKey forType(
+      MessageFactoryType type, String customPayloadSerializerClassName) {
+    return new MessageFactoryKey(type, customPayloadSerializerClassName);
+  }
+
+  public MessageFactoryType getType() {
+    return this.type;
+  }
+
+  public String getCustomPayloadSerializerClassName() {
```

Comment:
As it stands now, this method gets called from MessageTypeSerializer.Snapshot.writeSnapshot, for all types of message serializers, not just custom ones. For non-custom ones, a null is returned and this is the value that is written into the snapshot, after the MesssageFactoryType. So, the version 2 serialized format is always a MessageFactoryType value and then the custom payload serializer class name (string, which can be null), for all types of serializers.

The writeSnapshot method could be changed to only call getCustomPayloadSerializerClassName in the event that the serializer is custom, and then to either 1) force a null serializer class name to be written to the snapshot or 2) skip the write of the serializer class name altogether in the non-custom case. This would make the read/writeSnapshot code and testing a bit more complicated but would definitely be doable, would one of those options be better?

Reply:
That is a good point!
In that case, what do you think about making this method return an `Optional<String>`,
and version 2 would write an extra byte to indicate if the optional is present or not.

Something along the lines:
```
if (className.isPresent()) {
   out.writeBooolean(true);
   out.writeUtf8String(className.get());
} else {
   out.writeBooolean(false);
}
```

Reply:
For some reason, before I couldn't comment here, but now I can. Weird glitch. Anyway, I left a top-level comment/question for you on this one, just FYI.

Reply:
> 
> So the serialized format would already seem to be what you've requested! But if you'd prefer it get written without using `StringUtils.writeNullableString`, I'm happy to make that change.

Oh I see. My original thinking here was to prevent a future NPE by someone who would expect (mistakenly) that `messageFactoryKey.getCustomPayloadSerializerClassName()`  would always return a class name. Making the method return `Optional` would signal to the user of that class that it is not necessarily true.

And regarding `StringUtils` I'd rather to keep that logic local to the `Snapshot` class, to protect from accidental change in `StringUtils`.

Reply:
Great -- understood, I'll make that change and push a new version.

Reply:
These changes are in the code I just pushed, please let me know if it looks OK or if you need anything else. Thanks.